### PR TITLE
SDK generation from TypeSpec for CLU

### PR DESCRIPTION
**Pull request by Zach-King2**
Generating the SDK for CLU from TypeSpec

### CODE REVIEW

**Comment by Zach-King2** [comment]
> @m-nash I am currently working on migrating the CLU SDK to TypeSpec, but I have been running into a few issues when trying to generate code. When running the command "dotnet build /t:GenerateCode" I ran into this error:
> 
> image
> 
> Do you have any insight on how to solve this?
 This is the complete command output:

MSBuild version 17.6.3+07e294721 for .NET
  Determining projects to restore...
  All projects are up-to-date for restore.
  Using module powershell-yaml with version 0.4.7.
  Reading configuration from C:\Users\t-kingzach\Documents\SDK\azure-sdk-for-net\sdk\cognitivelanguage\Azure.AI.Language.Conversations\tsp-location.yaml
  Checking remote origin        https://github.com/Zach-King2/azure-sdk-for-net.git (fetch)
  Found git remote https://github.com/Zach-King2/azure-rest-api-specs.git
  from tsplocation.yaml 'repo' is: Zach-King2/azure-rest-api-specs
  Setting up sparse clone for Azure.AI.Language.Conversations at C:\Users\t-kingzach\Documents\SDK\sparse-spec\Azure.AI.Language.Conversations
  Updating sparse checkout file with directory:specification/cognitiveservices/Language.Conversations
  git checkout commit: e57839dc71bb80d16a8038eb3829ed4d4b2cc466
EXEC : warning : unrecognized pattern: 'specification/cognitiveservices/Language.Conversations' [C:\Users\t-kingzach\Documents\SDK\azure-sdk-for-net\sdk\cognitivelanguage\Azure.AI.Language.Conversations\src\Azure.AI.Language.Conversations.csproj]
EXEC : warning : disabling cone pattern matching [C:\Users\t-kingzach\Documents\SDK\azure-sdk-for-net\sdk\cognitivelanguage\Azure.AI.Language.Conversations\src\Azure.AI.Language.Conversations.csproj]
  HEAD is now at e57839dc71 added csharp emitter
  Copying spec from C:\Users\t-kingzach\Documents\SDK\sparse-spec\Azure.AI.Language.Conversations\specification\cognitiveservices\Language.Conversations to C:\Users\t-kingzach\Documents\SDK\azure-sdk-for-net\sdk\cognitivelanguage\Azure.AI.Language.Conversations\src/../TempTypeSpecFi
  les
  Using module powershell-yaml with version 0.4.7.
  Reading configuration from C:\Users\t-kingzach\Documents\SDK\azure-sdk-for-net\sdk\cognitivelanguage\Azure.AI.Language.Conversations\tsp-location.yaml
  Generating from C:\Users\t-kingzach\Documents\SDK\azure-sdk-for-net\sdk\cognitivelanguage\Azure.AI.Language.Conversations\TempTypeSpecFiles\Language.Conversations
  Copying package.json from C:\Users\t-kingzach\Documents\SDK\azure-sdk-for-net\eng\common\scripts\..\..\emitter-package.json
  
  added 295 packages, and audited 296 packages in 11s
  
  56 packages are looking for funding
    run `npm fund` for details
  
  found 0 vulnerabilities
  npx tsp compile C:\Users\t-kingzach\Documents\SDK\azure-sdk-for-net\sdk\cognitivelanguage\Azure.AI.Language.Conversations\TempTypeSpecFiles\Language.Conversations\client.tsp --emit @azure-tools/typespec-csharp --option @azure-tools/typespec-csharp.emitter-output-dir=C:\Users\t-kin
  gzach\Documents\SDK\azure-sdk-for-net\sdk\cognitivelanguage\Azure.AI.Language.Conversations/src
  TypeSpec compiler v0.46.0
  
  {"level":"info","message":"routes:1"}
  {"level":"info","message":"load operation: analyzeConversation, path:/language/:analyze-conversations "}
  {"level":"info","message":"Checking if C:/Users/t-kingzach/Documents/SDK/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/Azure.AI.Language.Conversations.csproj exists"}
  {"level":"info","message":"dotnet --roll-forward Major C:/Users/t-kingzach/Documents/SDK/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/TempTypeSpecFiles/Language.Conversations/node_modules/@autorest/csharp/AutoRest.CSharp.dll --project-path C:/Users/t-kin
  gzach/Documents/SDK/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/src   --clear-output-folder true"}
  Unhandled exception. System.NotSupportedException: No deserialization logic exists for ConversationalTaskResult
     at AutoRest.CSharp.Generation.Writers.JsonCodeWriterExtensions.GetDeserializeImplementationFormattable(TypeProvider implementation, FormattableString element, JsonSerializationOptions options) in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\Common\Generation\Writers\JsonC
  odeWriterExtensions.cs:line 816
     at AutoRest.CSharp.Generation.Writers.SerializationWriter.WriteJsonDeserialize(CodeWriter writer, TypeDeclarationOptions declaration, JsonObjectSerialization serialization) in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\Common\Generation\Writers\SerializationWriter.cs:li
  ne 205
     at AutoRest.CSharp.Generation.Writers.SerializationWriter.WriteObjectSerialization(CodeWriter writer, SerializableObjectType model, TypeDeclarationOptions declaration, JsonObjectSerialization jsonSerialization, XmlObjectSerialization xmlSerialization, Boolean isStruct, Boolean 
  includeSerializer, Boolean includeDeserializer) in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\Common\Generation\Writers\SerializationWriter.cs:line 100
     at AutoRest.CSharp.Generation.Writers.SerializationWriter.WriteObjectSerialization(CodeWriter writer, SerializableObjectType model) in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\Common\Generation\Writers\SerializationWriter.cs:line 45
     at AutoRest.CSharp.Generation.Writers.SerializationWriter.WriteSerialization(CodeWriter writer, TypeProvider schema) in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\Common\Generation\Writers\SerializationWriter.cs:line 35
     at AutoRest.CSharp.AutoRest.Plugins.LowLevelTarget.ExecuteAsync(GeneratedCodeWorkspace project, InputNamespace inputNamespace, SourceInputModel sourceInputModel, Boolean isTspInput) in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\LowLevel\AutoRest\LowLevelTarget.cs:line 3 
  2
     at AutoRest.CSharp.AutoRest.Plugins.CSharpGen.ExecuteAsync(InputNamespace rootNamespace) in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\Common\AutoRest\Plugins\CSharpGen.cs:line 59
     at AutoRest.CSharp.AutoRest.Communication.StandaloneGeneratorRunner.RunAsync(CommandLineOptions options) in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\Common\AutoRest\Communication\StandaloneGeneratorRunner.cs:line 51
     at AutoRest.CSharp.Program.Run(CommandLineOptions options) in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\Program.cs:line 47
     at AutoRest.CSharp.Program.<>c.<<Main>b__2_0>d.MoveNext() in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\Program.cs:line 29
  --- End of stack trace from previous location ---
     at AutoRest.CSharp.Program.Main(String[] args) in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\Program.cs:line 26
     at AutoRest.CSharp.Program.<Main>(String[] args)
  {"level":"info","message":"Command failed: dotnet --roll-forward Major C:/Users/t-kingzach/Documents/SDK/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/TempTypeSpecFiles/Language.Conversations/node_modules/@autorest/csharp/AutoRest.CSharp.dll --project-pat 
  h C:/Users/t-kingzach/Documents/SDK/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/src   --clear-output-folder true"}
  ExternalError: Emitter "@azure-tools/typespec-csharp" failed!
  File issue at https://github.com/Azure/autorest.csharp/issues

EXEC : error : Command failed: dotnet --roll-forward Major C:/Users/t-kingzach/Documents/SDK/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/TempTypeSpecFiles/Language.Conversations/node_modules/@autorest/csharp/AutoRest.CSharp.dll --project-path C:/Users/t-k 
ingzach/Documents/SDK/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/src   --clear-output-folder true [C:\Users\t-kingzach\Documents\SDK\azure-sdk-for-net\sdk\cognitivelanguage\Azure.AI.Language.Conversations\src\Azure.AI.Language.Conversations.csproj]       
      at checkExecSyncError (node:child_process:885:11)
      at execSync (node:child_process:957:15)
      at Object.$onEmit [as emitFunction] (file:///C:/Users/t-kingzach/Documents/SDK/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/TempTypeSpecFiles/Language.Conversations/node_modules/@azure-tools/typespec-csharp/dist/src/emitter.js:105:21)
      at async runEmitter (file:///C:/Users/t-kingzach/Documents/SDK/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/TempTypeSpecFiles/Language.Conversations/node_modules/@typespec/compiler/dist/src/core/program.js:563:13)
      at async compile (file:///C:/Users/t-kingzach/Documents/SDK/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/TempTypeSpecFiles/Language.Conversations/node_modules/@typespec/compiler/dist/src/core/program.js:257:9)
      at async Object.handler (file:///C:/Users/t-kingzach/Documents/SDK/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/TempTypeSpecFiles/Language.Conversations/node_modules/@typespec/compiler/dist/src/core/cli/cli.js:125:25)
      at checkExecSyncError (node:child_process:885:11)
      at execSync (node:child_process:957:15)
      at Object.$onEmit [as emitFunction] (file:///C:/Users/t-kingzach/Documents/SDK/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/TempTypeSpecFiles/Language.Conversations/node_modules/@azure-tools/typespec-csharp/dist/src/emitter.js:105:21)

      at async runEmitter (file:///C:/Users/t-kingzach/Documents/SDK/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/TempTypeSpecFiles/Language.Conversations/node_modules/@typespec/compiler/dist/src/core/program.js:563:13)
      at async compile (file:///C:/Users/t-kingzach/Documents/SDK/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/TempTypeSpecFiles/Language.Conversations/node_modules/@typespec/compiler/dist/src/core      at execSync (node:child_process:957:15)
      at Object.$onEmit [as emitFunction] (file:///C:/Users/t-kingzach/Documents/SDK/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/TempTypeSpecFiles/Language.Conversations/node_modules/@azure-tools/typespec-csharp/dist/src/emitter.js:105:21)
      at async runEmitter (file:///C:/Users/t-kingzach/Documents/SDK/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/TempTypeSpecFiles/Language.Conversations/node_modules/@typespec/compiler/dist/src/core/program.js:563:13)
      at async compile (file:///C:/Users/t-kingzach/Documents/SDK/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/TempTypeSpecFiles/Language.Conversations/node_modules/@typespec/compiler/dist/src/core/program.js:257:9)
      at async Object.handler (file:///C:/Users/t-kingzach/Documents/SDK/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/TempTypeSpecFiles/Language.Conversations/node_modules/@typespec/compiler/dist/src/core/cli/cli.js:125:25)
      at runEmitter (file:///C:/Users/t-kingzach/Documents/SDK/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/TempTypeSpecFiles/Language.Conversations/node_modules/@typespec/compiler/dist/src/core/program.js:583:19)
      at async compile (file:///C:/Users/t-kingzach/Documents/SDK/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/TempTypeSpecFiles/Language.Conversations/node_modules/@typespec/compiler/dist/src/core/program.js:257:9)
      at async Object.handler (file:///C:/Users/t-kingzach/Documents/SDK/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/TempTypeSpecFiles/Language.Conversations/node_modules/@typespec/compiler/dist/src/core/cli/cli.js:125:25)
C:\Users\t-kingzach\.nuget\packages\microsoft.azure.autorest.csharp\3.0.0-beta.20230717.1\buildMultiTargeting\Microsoft.Azure.AutoRest.CSharp.targets(60,5): error MSB3073: The command "pwsh -NoProfile -NonInteractive -executionpolicy Unrestricted -File C:\Users\t-kingzach\Documents\
SDK\azure-sdk-for-net\sdk\cognitivelanguage\Azure.AI.Language.Conversations\src/../../../../eng/common/scripts/TypeSpec-Project-Generate.ps1 C:\Users\t-kingzach\Documents\SDK\azure-sdk-for-net\sdk\cognitivelanguage\Azure.AI.Language.Conversations\src/..  " exited with code 1. [C:\Us
ers\t-kingzach\Documents\SDK\azure-sdk-for-net\sdk\cognitivelanguage\Azure.AI.Language.Conversations\src\Azure.AI.Language.Conversations.csproj]

Build FAILED.

EXEC : warning : unrecognized pattern: 'specification/cognitiveservices/Language.Conversations' [C:\Users\t-kingzach\Documents\SDK\azure-sdk-for-net\sdk\cognitivelanguage\Azure.AI.Language.Conversations\src\Azure.AI.Language.Conversations.csproj]
EXEC : warning : disabling cone pattern matching [C:\Users\t-kingzach\Documents\SDK\azure-sdk-for-net\sdk\cognitivelanguage\Azure.AI.Language.Conversations\src\Azure.AI.Language.Conversations.csproj]
EXEC : error : Command failed: dotnet --roll-forward Major C:/Users/t-kingzach/Documents/SDK/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/TempTypeSpecFiles/Language.Conversations/node_modules/@autorest/csharp/AutoRest.CSharp.dll --project-path C:/Users/t-k
ingzach/Documents/SDK/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/src   --clear-output-folder true [C:\Users\t-kingzach\Documents\SDK\azure-sdk-for-net\sdk\cognitivelanguage\Azure.AI.Language.Conversations\src\Azure.AI.Language.Conversations.csproj]
C:\Users\t-kingzach\.nuget\packages\microsoft.azure.autorest.csharp\3.0.0-beta.20230717.1\buildMultiTargeting\Microsoft.Azure.AutoRest.CSharp.targets(60,5): error MSB3073: The command "pwsh -NoProfile -NonInteractive -executionpolicy Unrestricted -File C:\Users\t-kingzach\Documents\
SDK\azure-sdk-for-net\sdk\cognitivelanguage\Azure.AI.Language.Conversations\src/../../../../eng/common/scripts/TypeSpec-Project-Generate.ps1 C:\Users\t-kingzach\Documents\SDK\azure-sdk-for-net\sdk\cognitivelanguage\Azure.AI.Language.Conversations\src/..  " exited with code 1. [C:\Us
ers\t-kingzach\Documents\SDK\azure-sdk-for-net\sdk\cognitivelanguage\Azure.AI.Language.Conversations\src\Azure.AI.Language.Conversations.csproj]
    2 Warning(s)
    2 Error(s)

File: sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/Generated/Configuration.json
Comment (heaths) [inline sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/Generated/Configuration.json:1]: @m-nash is this and the _tspCodeModel.json_ file supposed to be checked in? The latter seems more like an intermediate output file.

File: sdk/cognitivelanguage/Azure.AI.Language.Conversations/tsp-location.yaml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+directory: specification/cognitiveservices/Language.Conversations
+additionalDirectories: []
+commit: e57839dc71bb80d16a8038eb3829ed4d4b2cc466
+repo: Zach-King2/azure-rest-api-specs
```
Comment (heaths) [inline sdk/cognitivelanguage/Azure.AI.Language.Conversations/tsp-location.yaml:4]: This should be Azure/azure-rest-api-specs and use the current HEAD commit of your feature branch.

File: sdk/contosowidgetmanager/Azure.Contoso.WidgetManager/Azure.Contoso.WidgetManager.sln
Comment (heaths) [inline sdk/contosowidgetmanager/Azure.Contoso.WidgetManager/Azure.Contoso.WidgetManager.sln:1]: What's all the contoso stuff doing in this PR?

**Review by heaths — CHANGES_REQUESTED** [review]

File: sdk/contosowidgetmanager/Azure.Contoso.WidgetManager/Azure.Contoso.WidgetManager.sln
Comment (Zach-King2) [reply sdk/contosowidgetmanager/Azure.Contoso.WidgetManager/Azure.Contoso.WidgetManager.sln:1]: This was a mistake, I must've ran the incorrect command. I'll remove this now.

File: sdk/cognitivelanguage/Azure.AI.Language.Conversations/tsp-location.yaml (hunk 1 above)
Comment (Zach-King2) [reply sdk/cognitivelanguage/Azure.AI.Language.Conversations/tsp-location.yaml:4]: In order to make this change, I first have to merge the changes I made to the tspconfig.yaml file in order for the SDK generation to work. I'll create a PR for that now.

File: sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/Generated/Configuration.json
Comment (m-nash) [reply sdk/cognitivelanguage/Azure.AI.Language.Conversations/src/Generated/Configuration.json:1]: no, these aren't supposed to be checked in

**Comment by jsquire** [comment]
@chunyu3, @lirenhe, @m-nash, @AlexanderSher:  The failure looks generator-related and we could use some help diagnosing how to fix/work-around.   There is some time-sensitivity here, so your expertise would be greatly appreciated.

Would one of you please assist?

**Comment by chunyu3** [comment]
> No deserialization logic exists for ConversationalTaskResult

Hello @jsquire The root cause is the usage of Model `ConversationalTaskResult` is none.

**Comment by jsquire** [comment]
Thank you, @chunyu3!

**Comment by heaths** [comment]
@chunyu3 @m-nash that seems to have gotten @Zach-King2 further, but generated source is invalid. Take a look at [_.cs_](https://github.com/Azure/azure-sdk-for-net/pull/37820/files#diff-def3b09d37979c58d2284ec9389e1f7914645ec963984edb844c60fc372aa429) and [_object.cs_](https://github.com/Azure/azure-sdk-for-net/pull/37820/files#diff-434285e2dea4fbbc84b2cb599ffbc8922ab3a96947710363bc721133ae902722).

**Comment by chunyu3** [comment]
> @chunyu3 @m-nash that seems to have gotten @Zach-King2 further, but generated source is invalid. Take a look at [_.cs_](https://github.com/Azure/azure-sdk-for-net/pull/37820/files#diff-def3b09d37979c58d2284ec9389e1f7914645ec963984edb844c60fc372aa429) and [_object.cs_](https://github.com/Azure/azure-sdk-for-net/pull/37820/files#diff-434285e2dea4fbbc84b2cb599ffbc8922ab3a96947710363bc721133ae902722).
hello @Zach-King2  @heaths 
`.cs` file because there are anonymous models defined, see following. anonymous model is not supported, please provide a model name for it.

```
@doc("It is a wrap up of LUIS Generally Available response.")
model LuisTargetIntentResult extends TargetIntentResult {
  @doc("It is a wrap up of LUIS Generally Available response.")
  targetProjectKind: targetProjectKind.Luis;

  #suppress "@azure-tools/typespec-azure-core/bad-record-type" "LuisTargetIntentResult.result should not use ': Record<unknown>'. Use ': Record<string>' instead."
  @doc("The actual response from a LUIS Generally Available application.")
  result?: {};
}
```

`object.cs` file because a property `metadata` is defined as `object` type. Can you update to Record<T> or something else, we will discuss whether or not and how to support build-in type `object`
```
model KnowledgeBaseAnswer {
    @doc("List of questions associated with the answer.")
    questions?: Array<string>;

    @doc("Answer text.")
    answer?: string;

    @doc("Answer confidence score, value ranges from 0 to 1.")
    @minValue(0)
    @maxValue(1)
    @projectedName("client", "confidence")
    confidenceScore?: float64;

    @doc("ID of the QnA result.")
    @projectedName("client", "qnaId")
    id?: int32;

    @doc("Source of QnA result.")
    source?: string;

    #suppress "deprecated" " Deprecated: object is deprecated. Please use {} for an empty model, `Record<unknown>` for a record with unknown property types, `unknown[]` for an array."
    @doc("Metadata associated with the answer, useful to categorize or filter question answers.")
    metadata?: object;

    @doc("Dialog associated with Answer.")
    dialog?: KnowledgeBaseAnswerDialog;

    @doc("Answer span object of QnA with respect to user's question.")
    @projectedName("client", "shortAnswer")
    answerSpan?: AnswerSpan;
}
```

**Comment by chunyu3** [comment]
Hello @Zach-King2  @heaths 
There are some problems in the polymorphism definition.  Each polymorphism model should has distinct discriminator value.
In the spec, the base model has the same discriminator value with the derived model. see following example.

```
@doc("The base class of a conversation input task result.")
@discriminator("kind")
model AnalyzeConversationTaskResult {
  @doc("The base class of a conversation input task result.")
  kind: AnalyzeConversationTaskResultsKind.ConversationResult;
}

model ConversationalTaskResult extends AnalyzeConversationTaskResult {
  @doc("The results of a Conversation task.")
  kind: AnalyzeConversationTaskResultsKind.ConversationResult;

  @doc("Represents a conversation analysis response.")
  result: AnalyzeConversationResult;
}
```

You can define the base model without any discriminator as following
```
@doc("The base class of a conversation input task result.")
@discriminator("kind")
model AnalyzeConversationTaskResult {
  @doc("The base class of a conversation input task result.")
  kind: AnalyzeConversationTaskResultsKind;
}
```

**Comment by heaths** [comment]
> `.cs` file because there are anonymous models defined, see following. anonymous model is not supported, please provide a model name for it.

@chunyu3 @m-nash that's valid JSON, and what we were told to use in this case. Seems like a bug in the emitter if valid, recommended TypeSpec can't be represented. This is intentionally not modeled because the service wants to declare "any valid JSON object". The TypeSpec team is actually meeting to discuss this and might add something like a `json` type to represent this, but for now they told us to use `{}` instead of `unknown`, which can be any valid JSON type e.g., object, array, or primitive.

This is blocking. /cc @jsquire

**Comment by heaths** [comment]
@Zach-King2 or @quentinRobinson can you make the changes noted here: https://github.com/Azure/azure-sdk-for-net/pull/37820#issuecomment-1669216976

I'll file a bug regarding the `result?: {}`, but for now you might also make that `Record<unknown>`. It's not ideal but will hopefully unblock. I'm pushing discussions of how we might solve this long-term.

**Comment by heaths** [comment]
@chunyu3 @m-nash even the suppress method says to use `{}`:

> Deprecated: object is deprecated. Please use {} for an empty model, `Record<unknown>` for a record with unknown property types, `unknown[]` for an array.

While, long term, `{}` is probably not what we want, it's still valid TypeSpec. I'll file a bug, but I think we're going to want some new `Json` type we're [discussing in our teams channel](https://teams.microsoft.com/l/message/19:906c1efbbec54dc8949ac736633e6bdf@thread.skype/1690838455083?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&parentMessageId=1690838455083&teamName=Azure%20SDK&channelName=TypeSpec%20(Cadl)%20Discussion%20%F0%9F%90%AE&createdTime=1690838455083).

**Comment by heaths** [comment]
I opened https://github.com/Azure/autorest.csharp/issues/3653 to track the _.cs_ file and invalid source, but I still think you should change this to `Record<unknown>` for now. We'll hopefully land on something better soon.

**Comment by Zach-King2** [comment]
> @Zach-King2 or @quentinRobinson can you make the changes noted here: [#37820 (comment)](https://github.com/Azure/azure-sdk-for-net/pull/37820#issuecomment-1669216976)
> 
> I'll file a bug regarding the `result?: {}`, but for now you might also make that `Record<unknown>`. It's not ideal but will hopefully unblock. I'm pushing discussions of how we might solve this long-term.

@heaths I just created a PR for this: https://github.com/Azure/azure-rest-api-specs/pull/25250

**Comment by quentinRobinson** [comment]
@chunyu3 Hello, we were able to clean up a bunch of the errors we were having before but now we are running into a new error.  


```
{"level":"info","message":"dotnet --roll-forward Major C:/src/typespec/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/TempTypeSpecFiles/Language.Conversations/node_modules/@autorest/csharp/AutoRest.CSharp.dll --project-path C:/src/typespec/azure-sdk-for-net/sdk/cognitivelang
  uage/Azure.AI.Language.Conversations/src   --clear-output-folder true"}
  Unhandled exception. System.NotSupportedException: No deserialization logic exists for MultiLanguageConversationAnalysisInput
     at AutoRest.CSharp.Generation.Writers.JsonCodeWriterExtensions.GetDeserializeImplementationFormattable(TypeProvider implementation, FormattableString element, JsonSerializationOptions options) in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\Common\Generation\Writers\JsonCodeWriterExtensions
  .cs:line 833
     at AutoRest.CSharp.Generation.Writers.JsonCodeWriterExtensions.GetDeserializeValueFormattable(FormattableString element, CSharpType serializationType, SerializationFormat serializationFormat, JsonSerializationOptions serializationOptions) in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\Comm
  on\Generation\Writers\JsonCodeWriterExtensions.cs:line 691
     at AutoRest.CSharp.Generation.Writers.JsonCodeWriterExtensions.GetDeserializeValueFormattable(JsonValueSerialization serialization, FormattableString element) in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\Common\Generation\Writers\JsonCodeWriterExtensions.cs:line 671
     at AutoRest.CSharp.Generation.Writers.JsonCodeWriterExtensions.DeserializeValue(CodeWriter writer, JsonSerialization serialization, FormattableString element) in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\Common\Generation\Writers\JsonCodeWriterExtensions.cs:line 549
     at AutoRest.CSharp.Generation.Writers.JsonCodeWriterExtensions.DeserializeIntoObjectProperties(CodeWriter writer, IEnumerable`1 propertySerializations, FormattableString itemVariable, Dictionary`2 propertyVariables, Boolean shouldTreatEmptyStringAsNull) in D:\a\_work\1\s\autorest.csharp\src\AutoR
  est.CSharp\Common\Generation\Writers\JsonCodeWriterExtensions.cs:line 354
     at AutoRest.CSharp.Generation.Writers.JsonCodeWriterExtensions.WriteObjectInitialization(CodeWriter writer, JsonObjectSerialization serialization) in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\Common\Generation\Writers\JsonCodeWriterExtensions.cs:line 0
     at AutoRest.CSharp.Generation.Writers.SerializationWriter.WriteJsonDeserialize(CodeWriter writer, TypeDeclarationOptions declaration, JsonObjectSerialization serialization) in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\Common\Generation\Writers\SerializationWriter.cs:line 220
     at AutoRest.CSharp.Generation.Writers.SerializationWriter.WriteObjectSerialization(CodeWriter writer, SerializableObjectType model, TypeDeclarationOptions declaration, JsonObjectSerialization jsonSerialization, XmlObjectSerialization xmlSerialization, Boolean isStruct, Boolean includeSerializer,
  Boolean includeDeserializer) in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\Common\Generation\Writers\SerializationWriter.cs:line 100
     at AutoRest.CSharp.Generation.Writers.SerializationWriter.WriteObjectSerialization(CodeWriter writer, SerializableObjectType model) in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\Common\Generation\Writers\SerializationWriter.cs:line 45
     at AutoRest.CSharp.Generation.Writers.SerializationWriter.WriteSerialization(CodeWriter writer, TypeProvider schema) in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\Common\Generation\Writers\SerializationWriter.cs:line 35
     at AutoRest.CSharp.AutoRest.Plugins.LowLevelTarget.ExecuteAsync(GeneratedCodeWorkspace project, InputNamespace inputNamespace, SourceInputModel sourceInputModel, Boolean isTspInput) in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\LowLevel\AutoRest\LowLevelTarget.cs:line 33
     at AutoRest.CSharp.AutoRest.Plugins.CSharpGen.ExecuteAsync(InputNamespace rootNamespace) in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\Common\AutoRest\Plugins\CSharpGen.cs:line 62
     at AutoRest.CSharp.AutoRest.Communication.StandaloneGeneratorRunner.RunAsync(CommandLineOptions options) in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\Common\AutoRest\Communication\StandaloneGeneratorRunner.cs:line 51
     at AutoRest.CSharp.Program.Run(CommandLineOptions options) in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\Program.cs:line 47
     at AutoRest.CSharp.Program.<>c.<<Main>b__2_0>d.MoveNext() in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\Program.cs:line 29
  --- End of stack trace from previous location ---
     at AutoRest.CSharp.Program.Main(String[] args) in D:\a\_work\1\s\autorest.csharp\src\AutoRest.CSharp\Program.cs:line 26
     at AutoRest.CSharp.Program.<Main>(String[] args)
  {"level":"info","message":"Command failed: dotnet --roll-forward Major C:/src/typespec/azure-sdk-for-net/sdk/cognitivelanguage/Azure.AI.Language.Conversations/TempTypeSpecFiles/Language.Conversations/node_modules/@autorest/csharp/AutoRest.CSharp.dll --project-path C:/src/typespec/azure-sdk-for-net/s
  dk/cognitivelanguage/Azure.AI.Language.Conversations/src   --clear-output-folder true"}
  Emitter "@azure-tools/typespec-csharp" crashed! This is a bug.
  Please file an issue at https://github.com/Azure/autorest.csharp/issues
```


The error goes away if we set @convenientAPI(false)
```
#suppress "@azure-tools/typespec-azure-core/no-operation-id" "Operation ID is automatically generated by the OpenAPI emitters and should not normally be specified."
@added(Versions.v2023_04_01)
@doc("Analyzes the input conversation utterance.")
@pollingOperation(analyzeConversationJobStatus)
@convenientAPI(false)
op analyzeConversationSubmitJob is Operations.LongRunningResourceCreateWithServiceProvidedName<AnalyzeConversationJobsInput>;

#suppress "@azure-tools/typespec-azure-core/no-operation-id" "Operation ID is automatically generated by the OpenAPI emitters and should not normally be specified."
@added(Versions.v2023_04_01)
@doc("Cancel a long-running job for text analysis of conversations.")
@summary("Cancel a long-running Text Analysis conversations job")
@pollingOperation(analyzeConversationJobStatus)
@actionSeparator(":")
@convenientAPI(false)
@action("cancel")
op analyzeConversationCancelJob is Operations.LongRunningResourceDelete<AnalyzeConversationJobState, TypeSpec.Http.AcceptedResponse>;
```

**Pull request closed without merging**